components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsDate.java

```diff
@@ -0,0 +1,273 @@
+            } else {
+                this.rangeYears = null;
+            }
+        }
+        this.month = getNumericValueFromJSON(useFuzzyDate, JSON_MONTH_FIELDNAME, MIN_MONTH, MAX_MONTH);
+        this.day = getNumericValueFromJSON(useFuzzyDate, JSON_DAY_FIELDNAME, MIN_DAY, MAX_DAY);
+    }
+
+    /**
+     * Initialize from a string, which may either be an ISO date in the "yyyy-mm-dd" format,
+     * or a modified ISO string where the year is followed by an "s" to indicade that decade only is known.
+     * Also supported: "yyyy[s]-mm" and "yyyy[s]". Formally, the format is: "yyyy[s][-mm[-dd]]".
+     *
+     * @param phenoTipsDateString a date string.
+     */
+    public PhenoTipsDate(String phenoTipsDateString)
+    {
+        // treat null input as empty input
+        String useDateString = (phenoTipsDateString == null) ? "" : phenoTipsDateString;
+
+        // string: "1920s" or "1990" or "1990-05" or "1990-01-21" or "1990s-01-01"
+        Matcher m = DATE_STRING_REGEXP.matcher(useDateString);
+        if (m.find()) {
+            // note: month (#4) and day (#6) groups may be null, that is supported
+            this.year  = stringToIntegerInRange(m.group(1), null, null);
+            this.month = stringToIntegerInRange(m.group(4), MIN_MONTH, MAX_MONTH);
+            this.day   = stringToIntegerInRange(m.group(6), MIN_DAY, MAX_DAY);
+            if (m.group(2) != null) {
+                this.rangeYears = 10;
+            } else {
+                this.rangeYears = null;
+            }
+        } else {
+            this.rangeYears = null;
+            this.year = null;
+            this.month = null;
+            this.day = null;
+        }
+    }
+
+    // Accepts both numeric and string representations of a number.
+    // If range is given, returns null if the number is not in range
+    private Integer getNumericValueFromJSON(JSONObject json, String key, Integer validRangeMin, Integer validRangeMax)
+    {
+        if (!json.has(key)) {
+            return null;
+        }
+
+        // check if the key is an integer
+        int intValue = json.optInt(key, Integer.MIN_VALUE);
+        if (intValue != Integer.MIN_VALUE) {
+            if (valueInRange(intValue, validRangeMin, validRangeMax)) {
+                return intValue;
+            }
+            return null;
+        }
+
+        // check if the key is a string and try to parse the value from it
+        return stringToIntegerInRange(json.optString(key, ""), validRangeMin, validRangeMax);
+    }
+
+    // If range is given, returns null if the number is not in range
+    private Integer stringToIntegerInRange(String str, Integer validRangeMin, Integer validRangeMax)
+    {
+        try {
+            Integer value = Integer.parseInt(str);
+            if (!valueInRange(value, validRangeMin, validRangeMax)) {
+                return null;
+            }
+            return value;
+        } catch (Exception ex) {
+            return null;
```

Comment:
I personally got used to cunture that `return null;` is the last line of method, up to dev I guess